### PR TITLE
unref accounts in shink and pack when we're committed

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
there is a race with shrink and packing (and ancient appending) where we unref a dead account while preparing but then decide to not act on the slot. This addrefs the accounts. In the meantime, a zero lamport single ref account may have been prepared for marking dead. This race leads to refcount issues.

#### Summary of Changes
Only unref when we are committed to making the account dead. This is a much cleaner architecture. The index lookup performance we were worried about will be handled by the new scanning model that is rolling out soon.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -4306,6 +4304,35 @@ impl AccountsDb {
...
+    pub(crate) fn unref_shrunk_dead_accounts<'a>(
+        &self,
+        unrefed_pubkeys: impl Iterator<Item = &'a Pubkey>,
+        slot: Slot,
+    ) {
+        self.accounts_index.scan(
+            unrefed_pubkeys,
+            |pubkey, slot_refs, _entry| {
+                // pubkeys in `unrefed_pubkeys` were unref'd in `shrink_collect` above under the assumption that we would shrink everything.
+                // Since shrink is not occurring, we need to addref the pubkeys to get the system back to the prior state since the account still exists at this slot.
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs]: looks like this comment is incorrect! 
I removed it.

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -4338,43 +4364,24 @@ impl AccountsDb {
                 // clean needs to take care of this dead slot
                 self.accounts_index.add_uncleaned_roots([slot]);
             }
-            info!(
-                "Unexpected shrink for slot {} alive {} capacity {}, \
-                likely caused by a bug for calculating alive bytes.",
-                slot, shrink_collect.alive_total_bytes, shrink_collect.capacity
-            );
+            if !shrink_collect.all_are_zero_lamports {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: I think this belongs in a separate pr.

File: accounts-db/src/ancient_append_vecs.rs [hunk 3]
```diff
@@ -755,8 +737,15 @@ impl AccountsDb {
         let mut target_slots_sorted = Vec::with_capacity(len);
 
         // `shrink_collect` all accounts in the append vecs we want to combine.
-        // This also unrefs all dead accounts in those append vecs.
-        // This needs to serially iterate largest to smallest slot so that we unref older dead slots after we have visited the newer alive slots.
+        // The following comments are no longer correct?
```
Comment (jeffwashington) [inline accounts-db/src/ancient_append_vecs.rs]: i think these 2 comments need to be removed. ironically, we can now re-add the par_iter here if we want.

**Comment by jeffwashington** [comment]
@brooksprumo go ahead and turn your eyes on this when you can, please. I believe we have a race condition in master today that this fixes.

**Comment by HaoranYi** [comment]
Rebase to master. No functional change.

File: accounts-db/src/ancient_append_vecs.rs (hunk 3 above)
Comment (HaoranYi) [reply accounts-db/src/ancient_append_vecs.rs]: done!

File: accounts-db/src/accounts_db.rs (hunk 2 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah. moved to https://github.com/anza-xyz/agave/pull/2831

File: accounts-db/src/accounts_db.rs [hunk 4]
```diff
@@ -4309,6 +4307,34 @@ impl AccountsDb {
...
 
+    pub(crate) fn unref_shrunk_dead_accounts<'a>(
+        &self,
+        unrefed_pubkeys: impl Iterator<Item = &'a Pubkey>,
+        slot: Slot,
+    ) {
+        self.accounts_index.scan(
+            unrefed_pubkeys,
+            |pubkey, slot_refs, _entry| {
+                // pubkeys in `unrefed_pubkeys` were unref'd in `shrink_collect` above under the assumption that we would shrink everything.
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: doh. this comment is wrong. I copied this code probably. These keys are to be unref'd.

File: accounts-db/src/accounts_db.rs [hunk 5]
```diff
@@ -4309,6 +4307,34 @@ impl AccountsDb {
             .fetch_add(time.as_us(), Ordering::Relaxed);
     }
 
+    pub(crate) fn unref_shrunk_dead_accounts<'a>(
+        &self,
+        unrefed_pubkeys: impl Iterator<Item = &'a Pubkey>,
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: this should be `pubkeys` or `pubkeys_to_unref`

File: accounts-db/src/accounts_db.rs [hunk 6]
```diff
@@ -4341,43 +4367,14 @@ impl AccountsDb {
                 // clean needs to take care of this dead slot
                 self.accounts_index.add_uncleaned_roots([slot]);
             }
-            info!(
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: this should remain. the ohter pr just puts an `if` around it to quiet the noise.

File: accounts-db/src/accounts_db.rs (hunk 4 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah. fixed.

File: accounts-db/src/accounts_db.rs (hunk 5 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: fixed.

File: accounts-db/src/accounts_db.rs (hunk 6 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah. done.

File: accounts-db/src/accounts_db.rs [hunk 7]
```diff
@@ -4004,7 +4003,6 @@ impl AccountsDb {
                         // not exist in the re-written slot. Unref it to keep the index consistent with
                         // rewriting the storage entries.
                         unrefed_pubkeys.push(pubkey);
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: nit: Should this var (and field in LoadACcoutnsIndexForShrink) be renamed to something like `pubkeys_to_unref`, since we haven't unref'd them yet (but need to)?

This can be a separate PR too.

File: accounts-db/src/accounts_db.rs [hunk 8]
```diff
@@ -4004,7 +4003,6 @@ impl AccountsDb {
...
-                        result = AccountsIndexScanResult::Unref;
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:4007]: Should we keep these pubkeys in the in-mem index, since we are going to unref them soon?

iow:
```rust
result = AccountsIndexScanResult::KeepInMemory;
```

**Review by brooksprumo — COMMENTED** [review]
I still need to go through the tests. And likely do another pass.

File: accounts-db/src/accounts_db.rs (hunk 7 above)
Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Already mentioned here: https://github.com/anza-xyz/agave/pull/2806#discussion_r1743845388

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: done.

File: accounts-db/src/accounts_db.rs (hunk 8 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:4007]: well the plan is that we don't bring them into memory. if they are on disk, we will return None. If they are in memory, we'll look at them here. Since they're in memory, they'll remain in memory. If we are going to unref them, that means they have a ref from someone else, too, so they'll remain in memory anyway.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:4007]: Not necessarily. 
These are candidates. We may decide not to pack them later on.

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:4007]: Got it, yes that makes sense: if this account is not alive here, that means we have another account somewhere else and thus this index entry is dirty and will remain in memory.

File: accounts-db/src/accounts_db.rs [hunk 9]
```diff
@@ -4039,7 +4037,7 @@ impl AccountsDb {
 
         LoadAccountsIndexForShrink {
             alive_accounts,
-            unrefed_pubkeys,
+            unrefed_pubkeys: pubkeys_to_unref,
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: Can the `LoadAccountsIndexForShrink::unrefed_pubkeys` field be renamed too, please?

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah. renamed.

File: accounts-db/src/accounts_db.rs [hunk 10]
```diff
@@ -4159,7 +4158,7 @@ impl AccountsDb {
                 .for_each(|stored_accounts| {
                     let LoadAccountsIndexForShrink {
                         alive_accounts,
-                        mut unrefed_pubkeys,
+                        pubkeys_to_unref: mut unrefed_pubkeys,
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:4161]: nit: Here's another `unrefed_pubkeys` that can be renamed. (And the others in this function.)

It looks like `ShrinkCollect` has an `unrefed_pubkeys` field too. That can be renamed. IMO a subsequent PR is fine too.

File: accounts-db/src/accounts_db.rs [hunk 11]
```diff
@@ -4350,34 +4376,11 @@ impl AccountsDb {
...
-                    AccountsIndexScanResult::OnlyKeepInMemoryIfDirty
-                },
-                None,
-                true,
-                ScanFilter::All,
-            );
             return;
         }
 
+        self.unref_shrunk_dead_accounts(shrink_collect.unrefed_pubkeys.iter().cloned(), slot);
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:4382]: I am trying to think through the implications of moving where we call this function. I believe the current location most closely matches existing behavior. 

At the other extreme, we could move this to after we remove the old storage. I think that's probably too far; I imagine we want to decrement the ref count before removing the storage, as that'd be the order I'd check for things.

Would we want to move decrementing the ref count until after the new storage was written? I'm not sure. If transaction processing tries to load an account mid-shrink—*after* decrementing the ref count, but *before* writing the new storage—doesn't the get_account_accessor retry logic handle that case too? I think we're safe. Probably better to get the new storage in that case.

Just thinking out loud here :)

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:4382]: once we pass any early returns, the unref will happen and the storage will be rewritten unless we panic or assert or oom. Nobody cares about refcount except clean, shrink, index, ancient pack. Only shrink or ancient pack are running at this time. Fg processes won't care what refcount is. These entries are already not in the slot list, so no fg process will ever know to look for this account in this slot.

**Review by brooksprumo — APPROVED** [review]
Code looks good to me. The rename nits can be handled here in this PR, or in a follow up.
